<details>
<summary>Pre-Commit Test failed! Click here</summary>

```
[INFO] Initializing environment for git://github.com/Lucas-C/pre-commit-hooks.
[INFO] Initializing environment for git://github.com/pre-commit/pre-commit-hooks.
[INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Initializing environment for https://github.com/adrienverge/yamllint.git.
[INFO] Installing environment for git://github.com/Lucas-C/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for git://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/adrienverge/yamllint.git.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
Tabs remover.............................................................Passed
Trim Trailing Whitespace.................................................Failed
- hook id: trailing-whitespace
- exit code: 1
- files were modified by this hook

Fixing docs/manage_your_app_secrets.md
Fixing docs/admin/add_new_cluster_spec.md
Fixing docs/admin/add_namespace_to_cluster.md

Check for merge conflicts................................................Passed
Fix End of Files.........................................................Failed
- hook id: end-of-file-fixer
- exit code: 1
- files were modified by this hook

Fixing docs/admin/add_namespace_to_cluster.md
Fixing README.md
Fixing docs/inclusions_explained.md

Check for added large files..............................................Passed
Check for case conflicts.................................................Passed
Check JSON...........................................(no files to check)Skipped
Check for broken symlinks............................(no files to check)Skipped
Detect Private Key.......................................................Passed
Fix End of Files.........................................................Passed
Trim Trailing Whitespace.................................................Passed
yamllint.................................................................Failed
- hook id: yamllint
- exit code: 1

manifests/overlays/dev/rbac/argocd-manager-role.yaml
  1:1       warning  missing document start "---"  (document-start)
  6:1       error    wrong indentation: expected 2 but found 0  (indentation)
  7:3       error    wrong indentation: expected 4 but found 2  (indentation)
  9:3       error    wrong indentation: expected 4 but found 2  (indentation)
  11:3      error    wrong indentation: expected 4 but found 2  (indentation)
  13:3      error    wrong indentation: expected 4 but found 2  (indentation)
  15:3      error    wrong indentation: expected 4 but found 2  (indentation)

manifests/base/deployments/argocd-repo-server_patch.yaml
  1:1       warning  missing document start "---"  (document-start)
  9:7       error    wrong indentation: expected 8 but found 6  (indentation)
  12:9      error    wrong indentation: expected 10 but found 8  (indentation)
  15:9      error    wrong indentation: expected 10 but found 8  (indentation)
  22:7      error    wrong indentation: expected 8 but found 6  (indentation)

manifests/crds/kustomization.yaml
  1:1       warning  missing document start "---"  (document-start)
  4:1       error    wrong indentation: expected 2 but found 0  (indentation)

manifests/base/services/argocd_metrics_service_patch.yaml
  1:1       warning  missing document start "---"  (document-start)

manifests/crds/aggregate-to-admin.yaml
  1:1       warning  missing document start "---"  (document-start)
  8:1       error    wrong indentation: expected 2 but found 0  (indentation)
  9:3       error    wrong indentation: expected 4 but found 2  (indentation)
  11:3      error    wrong indentation: expected 4 but found 2  (indentation)
  14:3      error    wrong indentation: expected 4 but found 2  (indentation)
  30:1      error    wrong indentation: expected 2 but found 0  (indentation)
  31:3      error    wrong indentation: expected 4 but found 2  (indentation)
  33:3      error    wrong indentation: expected 4 but found 2  (indentation)
  35:3      error    wrong indentation: expected 4 but found 2  (indentation)

manifests/overlays/dev/projects/projects.yaml
  1:1       warning  missing document start "---"  (document-start)
  7:3       error    wrong indentation: expected 4 but found 2  (indentation)
  10:3      error    wrong indentation: expected 4 but found 2  (indentation)

manifests/overlays/dev/secrets/auth/argocd-secret.enc.yaml
  1:1       warning  missing document start "---"  (document-start)
  9:81      error    line too long (207 > 80 characters)  (line-length)
  10:81     error    line too long (160 > 80 characters)  (line-length)
  11:81     error    line too long (1432 > 80 characters)  (line-length)
  18:81     error    line too long (288 > 80 characters)  (line-length)
  20:8      error    too many spaces after hyphen  (hyphens)
  20:5      error    wrong indentation: expected 8 but found 4  (indentation)

manifests/overlays/dev/projects/kustomization.yaml
  1:1       warning  missing document start "---"  (document-start)
  4:1       error    wrong indentation: expected 2 but found 0  (indentation)

manifests/base/services/argocd_server_metrics_service_patch.yaml
  1:1       warning  missing document start "---"  (document-start)

manifests/overlays/dev/rbac/argocd-manager-role-binding.yaml
  1:1       warning  missing document start "---"  (document-start)
  10:1      error    wrong indentation: expected 2 but found 0  (indentation)

manifests/overlays/dev/kustomization.yaml
  1:1       warning  missing document start "---"  (document-start)
  5:1       error    wrong indentation: expected 2 but found 0  (indentation)
  11:1      error    wrong indentation: expected 2 but found 0  (indentation)
  25:1      error    wrong indentation: expected 2 but found 0  (indentation)
  29:1      error    wrong indentation: expected 2 but found 0  (indentation)

manifests/base/serviceaccounts/argocd-manager.yaml
  1:1       warning  missing document start "---"  (document-start)

manifests/base/kustomization.yaml
  1:1       warning  missing document start "---"  (document-start)
  4:1       error    wrong indentation: expected 2 but found 0  (indentation)
  8:1       error    wrong indentation: expected 2 but found 0  (indentation)

manifests/overlays/dev/applications/dev_app/kustomization.yaml
  1:1       warning  missing document start "---"  (document-start)
  4:1       error    wrong indentation: expected 2 but found 0  (indentation)

manifests/overlays/dev/secrets/gpg/secret-generator.yaml
  1:1       warning  missing document start "---"  (document-start)
  6:1       error    wrong indentation: expected 2 but found 0  (indentation)

manifests/overlays/dev/serviceaccounts/argocd-dex-server.yaml
  1:1       warning  missing document start "---"  (document-start)
  10:81     error    line too long (94 > 80 characters)  (line-length)

manifests/overlays/dev/applications/kustomization.yaml
  1:1       warning  missing document start "---"  (document-start)
  4:1       error    wrong indentation: expected 2 but found 0  (indentation)

examples/argocd-manager-rolebinding.yaml
  1:1       warning  missing document start "---"  (document-start)

manifests/overlays/dev/secrets/clusters/secret-generator.yaml
  1:1       warning  missing document start "---"  (document-start)
  6:1       error    wrong indentation: expected 2 but found 0  (indentation)

manifests/overlays/dev/secrets/auth/srcops.enc.yaml
  1:1       warning  missing document start "---"  (document-start)
  3:81      error    line too long (194 > 80 characters)  (line-length)
  4:81      error    line too long (137 > 80 characters)  (line-length)
  15:81     error    line too long (288 > 80 characters)  (line-length)
  17:8      error    too many spaces after hyphen  (hyphens)
  17:5      error    wrong indentation: expected 8 but found 4  (indentation)

manifests/base/deployments/argocd-server_patch.yaml
  1:1       warning  missing document start "---"  (document-start)
  9:7       error    wrong indentation: expected 8 but found 6  (indentation)
  11:9      error    wrong indentation: expected 10 but found 8  (indentation)

manifests/overlays/dev/applications/dev_app/dev-app.yaml
  1:1       warning  missing document start "---"  (document-start)
  16:5      error    wrong indentation: expected 6 but found 4  (indentation)

manifests/overlays/dev/secrets/clusters/dev.cluster.example.enc.yaml
  1:1       warning  missing document start "---"  (document-start)
  11:81     error    line too long (133 > 80 characters)  (line-length)
  12:81     error    line too long (227 > 80 characters)  (line-length)
  13:81     error    line too long (163 > 80 characters)  (line-length)
  14:81     error    line too long (135 > 80 characters)  (line-length)
  21:81     error    line too long (288 > 80 characters)  (line-length)
  23:8      error    too many spaces after hyphen  (hyphens)
  23:5      error    wrong indentation: expected 8 but found 4  (indentation)

manifests/overlays/dev/secrets/auth/secret-generator.yaml
  1:1       warning  missing document start "---"  (document-start)
  6:1       error    wrong indentation: expected 2 but found 0  (indentation)

manifests/overlays/dev/secrets/auth/dex-server-sa.enc.yaml
  1:1       warning  missing document start "---"  (document-start)
  15:81     error    line too long (288 > 80 characters)  (line-length)
  17:8      error    too many spaces after hyphen  (hyphens)
  17:5      error    wrong indentation: expected 8 but found 4  (indentation)

manifests/base/routes/argocd_server_route.yaml
  1:1       warning  missing document start "---"  (document-start)

.aicoe-ci.yaml
  1:1       warning  missing document start "---"  (document-start)

manifests/overlays/dev/secrets/gpg/secret.enc.yaml
  1:1       warning  missing document start "---"  (document-start)
  6:81      error    line too long (4696 > 80 characters)  (line-length)
  13:81     error    line too long (288 > 80 characters)  (line-length)
  15:8      error    too many spaces after hyphen  (hyphens)
  15:5      error    wrong indentation: expected 8 but found 4  (indentation)
```
</details>

--- a/README.md
+++ b/README.md
@@ -10,13 +10,13 @@ KSOPS 2.1.2+
 Ensure you have the key to decrypt secrets. Reach out to members of the Data Hub team for access.
 
 ### GPG Key access
- 
+
 This repo encrypts secrets using a dev test key, you can find the test key in examples/key.asc folder.
 
 ```
 $ cat examples/key.asc > base64 -d
 ```
 
-You will need to import this key to be able to decrypt the contents of the secrets using sops. 
+You will need to import this key to be able to decrypt the contents of the secrets using sops.
 
-Do NOT use this gpg key for prod purposes. 
+Do NOT use this gpg key for prod purposes.

--- a/docs/admin/add_namespace_to_cluster.md
+++ b/docs/admin/add_namespace_to_cluster.md
@@ -1,0 +1,33 @@
+# Add namespace to cluster
+
+## Prerequisites
+
+* sops 3.6+
+* sops access
+
+## Instructions
+
+Namespaces are added to ArgoCD by altering the corresponding cluster spec. Cluster specs are defined within the `/manifests/overlays/<env>/secrets/clusters` folder.
+
+Open the file in the sops editor, for example if updating the cluster spec `dev.cluster.example.enc.yaml` in `dev` you would execute:
+
+```bash
+# From repo root
+$ target_env=dev
+$ cd manifests/overlays/$target_env/secrets/clusters
+$ sops dev.cluster.example.enc.yaml
+```
+
+This should open the decrypted form of the cluster spec. Update the namespace field by appending your namespace (comma-separated, no spaces, if there are multiple namespaces).
+
+```yaml
+...
+ 10 stringData:
+ 11     name: dev-cluster-spec
+ 12     config: ...
+ 14     namespaces: namespace-1,namespace-2 # Update this field by appending the new namespace, no whitespaces between commas
+ 15     server: ...
+```
+
+# Resources
+More info on cluster configuration can be found [here](https://argoproj.github.io/argo-cd/operator-manual/declarative-setup/#clusters).

--- a/docs/admin/add_new_cluster_spec.md
+++ b/docs/admin/add_new_cluster_spec.md
@@ -1,0 +1,67 @@
+# Adding a new cluster spec
+
+## Prerequisites
+
+* sops 3.6+
+* sops access
+
+## Instructions
+
+ArgoCD will need a service account present on the cluster for deployments. Where the SA is located is irrelevant, though it's advised to have it be located in its own independent namespace. For consistency name this service account `argocd-manager`.
+
+This workflow may look like this:
+
+```bash
+oc login <your_cluster>
+oc new-project argocd-manager
+oc create sa argocd-manager
+```
+
+Get the token for this SA
+
+```bash
+SA_TOKEN=`oc sa get-token argocd-manager -n argocd-manager`
+```
+
+Store the cluster specs in `/manifests/overlays/<env>/secrets/clusters` folder.
+
+Create the cluster spec:
+```yaml
+# /manifests/overlays/dev/secrets/clusters/dev.cluster.example.yaml
+apiVersion: v1
+kind: Secret
+metadata:
+    name: dev-cluster-spec
+    labels:
+        argocd.argoproj.io/secret-type: cluster
+    annotations:
+        managed-by: argocd.argoproj.io
+type: Opaque
+stringData:
+    name:  dev.cluster.com
+    config: |
+        {"bearerToken": ${SA_TOKEN}, "tlsClientConfig": {"insecure": true}}
+    namespaces: namespace_1,namespace_2
+    server: https://dev.cluster.uri.com:44
+```
+
+Let's go over what some of the fields in the `stringData` field refer to:
+
+- `name`: Name for this cluster, appears in the ArgoCD UI
+
+- `config`: The token goes here, replace the contents of ${SA_TOKEN} with the one retrieved earlier.
+
+- `namespace`: List of namespaces ArgoCD has permissions to deploy to, comma-separated, no whitespace
+
+- `server`: Cluster api server hostname, can be retrieved by running `oc whoami --show-server`
+
+SOPS encrypt the manifest and store it in the `clusters` directory.
+```
+$ target_env=dev
+$ cd manifests/overlays/$target_env/secrets/clusters
+$ sops -e dev.cluster.example.yaml > dev.cluster.example.enc.yaml
+# Delete the cluster_spec.yaml (decrypted version)
+$ rm dev.cluster.example.yaml
+```
+
+> Note: DO NOT submit a PR with the decrypted cluster spec secret.

--- a/docs/admin/update_gpg_key.md
+++ b/docs/admin/update_gpg_key.md
@@ -1,0 +1,24 @@
+# Update gpg key
+
+## Prerequisites
+
+* sops 3.6+
+
+## Instructions
+
+Export the key
+
+```bash
+$ gpg --export-secret-keys "${KEY_ID}" | base64 > private.asc
+```
+
+```bash
+# From the repo root
+$ target_env=dev
+$ cd manifests/overlays/$target_env/secrets/gpg
+$ sops secret.enc.yaml
+```
+
+Copy the contents of `private.asc` into the `private.key` field.
+
+Save the file, exit. Commit and make a PR.

--- a/docs/create_argocd_application_manifest.md
+++ b/docs/create_argocd_application_manifest.md
@@ -1,0 +1,97 @@
+# Application Management
+
+While ArgoCD allows you to create ArgoCD applications via the UI and CLI, we recommend that all applications be
+created [declaratively](https://argoproj.github.io/argo-cd/operator-manual/declarative-setup/#applications).
+
+This allows you to easily restore your applications should the need arise.
+
+## Pre-requisites
+- Kustomize version 3.8+
+
+## Steps for creating an application
+For your application to show up to ArgoCD you need to do 2 things:
+1. Create the Application yaml in the appropriate path in a fork
+2. Submit a PR to the base repository
+
+These steps are outlined in detail below:
+
+### Step 1. Create the Application Yaml
+
+Clone the repo and `cd` into where applications are stored:
+
+```bash
+$ target_env=dev
+$ cd /manifests/overlays/$target_env/applications
+```
+
+If your team folder does not exist, create it:
+
+```bash
+$ mkdir example_team && cd example_team
+$ kustomize create
+```
+
+Let's create a sample application called `example-app`.
+
+```yaml
+# /manifests/overlays/dev/applications/example_team/example_app.yaml
+
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: example_app
+spec:
+  destination:
+    namespace: example_namespace
+    server: http://example_server
+  project: example_project
+  source:
+    path: path/to/kustomization
+    repoURL: http://path-to-example-repo
+    targetRevision: HEAD
+  syncPolicy:
+    automated:
+      prune: true
+      selfHeal: true
+    syncOptions:
+    - Validate=false
+```
+This is a basic minimal example application. For additional fields see [here](https://argoproj.github.io/argo-cd/operator-manual/application.yaml).
+
+Let's go over what some of the fields in the `example_app.yaml` refer to:
+
+- `metadata.name`: the name of the `Application` resource as well as the name as it appears on the ui.
+
+- `spec.project`: the ArgoCD `Project` to which this `Application` belongs, ensure that this `Project` exists in `manifests/overlays/<target_env>/projects`.
+
+- `spec.destination.namespace`: the target namespace for this `Application`'s deployment, ensure that this namespace exists in `manifests/overlays/<target_env>/secrets/clusters` for the appropriate cluster/server.
+
+- `spec.destination.cluster`: the target cluster server name for this `Application`'s deployment, ensure this server exists in  `manifests/overlays/<target_env>/secrets/clusters` for the appropriate cluster.
+
+- `spec.source.path`: path to the Kustomization.yaml file relative to the repo's root.
+
+- `spec.source.repoURL`: the repository holding the `Application`'s desired state, ensure this repo exists within `argo_cm` located in `manifests/overlays/<target_env>/configs/argo_cm/repositories`.
+
+> NOTE: You may need to disable schema validation if your deployments are failing. This is due to the ArgoCD api validator having a very strict api spec.
+
+Once you are done add the application yaml to the Kustomization file by running the following _from the repository root directory_.
+
+```
+$ target_env=dev
+$ cd /manifests/overlays/$target_env/applications/example_team
+$ kustomize edit add resource example_app.yaml
+```
+
+### Step 2. Make a Pull request
+Commit your changes and submit a pr to your continuous deployment repository.
+
+#### If your application exists in ArgoCD but not on VCS
+Getting the manifests for all your applications is easy to do with the argocd cli.
+
+```bash
+# Login via cli using sso
+$ argocd --insecure --grpc-web login ${ARGOCD_ROUTE}:443 --sso
+
+# Get the application resource details
+$ argocd app get ${APP_NAME} -o yaml
+```

--- a/docs/get_argocd_to_manage_your_app.md
+++ b/docs/get_argocd_to_manage_your_app.md
@@ -1,0 +1,23 @@
+# Get ArgoCD to manage your Application
+
+When migrating an application's deployment to be managed by ArgoCD use the following checklist to verify your process.
+
+- Ensure your application manifests can be built using Kustomize.
+- If using secrets, make sure to include the .sops.yaml file in your repository.
+    - See [here](manage_your_app_secrets.md) for more info.
+- Create the role granting access to namespace.
+    - See [here](give_argocd_access_to_your_project.md) for more info.
+    - This role should be tracked in your application manifest repository.
+
+The following items require a PR:
+
+- Ensure the application repository is added in the `repository` file in `/manifests/overlays/<target_env>/configs/argo_cm/repositories`.
+- Ensure that all OCP resources that will be managed by ArgoCD on this cluster are included in the `inclusions` list in `/manifests/overlays/<target_env>/configs/argo_cm/resource.inclusions`.
+    - See [here](inclusions_explained.md) for more info.
+- Create the ArgoCD Application manifest
+    - See [here](create_argocd_application_manifest.md) for more info.
+
+The following items require a PR with sops access:
+
+- Ensure your namespace exists in your cluster's spec see [here](admin/add_new_cluster_spec.md) for details.
+- If you are switching between ArgoCD managed namespaces, and that namespace was deleted in OCP, then ensure it's also removed from your cluster's credentials found here `/manifests/overlays/<target_env>/secrets/clusters`.

--- a/docs/give_argocd_access_to_your_project.md
+++ b/docs/give_argocd_access_to_your_project.md
@@ -1,0 +1,29 @@
+# Give ArgoCD access to your project
+
+ArgoCD uses an SA named `argocd-manager` to deploy resources to another cluster/namespace. These SAs need access to the resources it will be deploying, this is done via roles and rolebindings.
+
+In your namespace, you will need to deploy a rolebinding like the one below:
+
+```yaml
+apiVersion: authorization.openshift.io/v1
+kind: RoleBinding
+metadata:
+  name: argocd-manager-rolebinding
+  namespace: <application_namespace>
+roleRef:
+  name: <role>
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+subjects:
+  - kind: ServiceAccount
+    name: argocd-manager
+    namespace: <sa_namespace>
+```
+
+Fill out `application_namespace`, `role`, and `sa_namespace`.
+
+> **`application_namespace`**: This is your project namespace.
+>
+> **`sa_namespace`**: The namespace where the service account resides.
+>
+> **`role`**: must be a project `admin` role.

--- a/docs/inclusions_explained.md
+++ b/docs/inclusions_explained.md
@@ -1,0 +1,18 @@
+# Inclusions explained
+
+It is likely that your team does not have `get` access to all namespace scoped resources.
+This can be an issue when deploying apps to a namespace in a cluster, because ArgoCD will
+attempt to discover all namespace scoped resourced and be denied. To avoid this, we limit
+ArgoCD to discover the resources that are available to project admins, these should be added
+under the `resource.inclusions` ArgoCD configurations in `manifests/overlays/<target_env>/configs/argo_cm/resource.inclusions`.
+
+If your application contains resources that a project `admin` does not have permissions
+to list/edit then you can request that a cluster admin deploy aggregated roles to add
+such permissions. See here for an [example](https://github.com/argoproj/argo-events/blob/master/manifests/cluster-install/rbac/argo-events-aggregate-to-admin.yaml).
+
+Once having done so, you can make a PR with these resources added onto the `resource.inclusions`
+list.
+
+## Resources
+
+Read more about exclusion/inclusions [here](https://argoproj.github.io/argo-cd/operator-manual/declarative-setup/#resource-exclusioninclusion).

--- a/docs/index.md
+++ b/docs/index.md
@@ -1,0 +1,2 @@
+Here you will find a series of docs that outline various procedures and how-tos
+when interacting with ArgoCD.

--- a/docs/manage_your_app_secrets.md
+++ b/docs/manage_your_app_secrets.md
@@ -1,0 +1,127 @@
+# Secret Management
+
+Secret management is handled using the KSOPs plugin. Use [sops](https://github.com/mozilla/sops) to encrypt your secrets in vcs.
+
+## Overview: KSOPs
+`KSOPS`, or kustomize-SOPS, is a [kustomize](https://github.com/kubernetes-sigs/kustomize/) plugin for SOPS encrypted resources. `KSOPS` can be used to decrypt any Kubernetes resource, but is most commonly used to decrypt encrypted Kubernetes Secrets and ConfigMaps. As a [kustomize](https://github.com/kubernetes-sigs/kustomize/) plugin, `KSOPS` allows you to manage, build, and apply encrypted manifests the same way you manage the rest of your Kubernetes manifests.
+
+## Requirements
+- [Go](https://github.com/golang/go)
+- [kustomize](https://github.com/kubernetes-sigs/kustomize/) built with Go (See [details below](#kustomize-go-plugin-caveats))
+- [SOPS](https://github.com/mozilla/sops)
+- gpg
+
+See [versions](versions.md) to download the appropriate version of SOPS, Kustomize, and KSOPS.
+
+### 0. Verify Requirements
+Before continuing, verify your installation of [Go](https://github.com/golang/go), [SOPS](https://github.com/mozilla/sops), and `gpg`. Below are a few non-comprehensive commands to quickly check your installations:
+
+```bash
+# Verify that the latest version of Go is installed i.e. v1.13 and above
+go version
+
+# Verify that your $GOPATH is set
+go env
+
+# Verify SOPS is installed
+sops --version
+
+# Verify gpg is installed
+gpg --help
+```
+
+### 1. Download KSOPS
+
+```bash
+# export GO111MODULE=on
+go get -u github.com/viaduct-ai/kustomize-sops
+# cd into the root directory
+cd $GOPATH/src/github.com/viaduct-ai/kustomize-sops
+```
+
+### 2. Install (or Reinstall) the Latest kustomize via Go
+
+```bash
+# KSOPS is built with latest kustomize
+# If you want to change versions, update the installation script with your desired version and make sure to check that the KSOPS tests still pass
+# If you want to change versions below kustomize v3.3.0, use the KSOPS v1.0 or go-1.12 release!
+make kustomize
+```
+
+### 3. Setup kustomize Plugin Path
+
+```bash
+# Don't forget to define XDG_CONFIG_HOME in your .bashrc/.zshrc
+echo "export XDG_CONFIG_HOME=\$HOME/.config" >> $HOME/.bashrc
+source $HOME/.bashrc
+```
+
+### 4. Build and Install KSOPS Plugin
+
+```bash
+make install
+```
+
+### 5. Configure SOPS via .sops.yaml
+
+`KSOPS` relies on the `SOPS` creation rules defined in `.sops.yaml`. To make encrypted secrets more readable, we suggest using the following encryption regex to only encrypt `data` and `stringData` values. This leaves non-sensitive fields, like the secret's name, unencrypted and human readable.
+
+You will have to modify `.sops.yaml` if you want to use your key management service by providing the correct gpg fingerprint. You can customize this file according to the type of secrets you want to encrypt.
+
+```yaml
+creation_rules:
+  - encrypted_regex: '^(data|stringData)$'
+    # Specify kms/pgp/etc encryption key
+    pgp: '<gpg-fingerprint>'
+    # Optionally you can configure to use a providers key store
+    # kms: XXXXXX
+    # gcp_kms: XXXXXX
+```
+
+### 6. Create a Resource
+
+```bash
+# Create a local Kubernetes Secret
+cat <<EOF > secret.yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: mysecret
+type: Opaque
+data:
+  username: YWRtaW4=
+  password: MWYyZDFlMmU2N2Rm
+EOF
+```
+
+### 7. Encrypt the Resource
+
+```bash
+# Encrypt with SOPS CLI
+# Specify SOPS configuration in .sops.yaml
+sops -e secret.yaml > secret.enc.yaml
+```
+
+### 8. Define KSOPS kustomize Generator
+```bash
+# Create a local Kubernetes Secret
+cat <<EOF > secret-generator.yaml
+apiVersion: viaduct.ai/v1
+kind: ksops
+metadata:
+  # Specify a name
+  name: example-secret-generator
+files:
+  - ./secret.enc.yaml
+EOF
+```
+
+### 9. Create the kustomization.yaml
+[Read about kustomize plugins](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/plugins/README.md)
+
+```bash
+cat <<EOF > kustomization.yaml
+generators:
+  - ./secret-generator.yaml
+EOF
+```

--- a/docs/setup_argocd_dev_enviornment.md
+++ b/docs/setup_argocd_dev_enviornment.md
@@ -1,0 +1,52 @@
+# Deploying a development environment
+
+## Prequisites
+* An OCP 4.x Development cluster
+* Must have cluster admin
+
+## Instructions
+
+Create the project `aicoe-argocd-dev` and `argocd-test`. The latter wll be used
+for deploying a dev application via ArgoCD.
+
+```bash
+oc new-project argocd-test
+oc new-project aicoe-argocd-dev
+```
+
+Deploy ArgoCD
+```bash
+git clone git@github.com:operate-first/continuous-deployment.git
+cd continuous-deployment
+
+# Deploy Cluser objects
+kustomize build manifests/crds --enable_alpha_plugins | oc apply -f -
+
+# Deploy Non Cluster objects
+kustomize build manifests/overlays/dev --enable_alpha_plugins | oc apply -f -
+```
+
+## Configure Auth
+Once deployed, there are some additional configurations, run this script:
+```bash
+examples/configure_development.sh
+```
+Feel free to look inside the script for detailed comments on what configurations are applied.
+
+## Cleanup
+Run the following commands to clean up your environment.
+
+```
+kustomize build manifests/overlays/dev --enable_alpha_plugins | oc delete -f -
+kustomize build manifests/crds --enable_alpha_plugins | oc delete -f -
+oc delete group dev
+oc delete project argocd-test
+oc delete project aicoe-argocd-dev
+```
+
+You may ignore the following error when removing secrets:
+
+```
+Error from server (NotFound): error when deleting "STDIN": secrets "argocd-dex-server-oauth-token" not found
+Error from server (NotFound): error when deleting "STDIN": secrets "dev-cluster-spec" not found
+```

--- a/docs/versions.md
+++ b/docs/versions.md
@@ -1,0 +1,13 @@
+# Versions
+
+ArgoCD: 1.6.1
+
+KSOPs: 2.1.2
+
+Kustomize: 3.8.0
+
+SOPS: 3.6.0+
+
+The KSOPS and Kustomize versions refer to the ones provisioned with ArgoCD.
+
+Kustomize versions can be adjusted manually using [customized versions](https://argoproj.github.io/argo-cd/user-guide/kustomize/#custom-kustomize-versions).

